File: use-cases/merchant-fulfillment/code/java/src/main/java/lambda/InventoryCheckHandler.java
package lambda;

import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.google.gson.Gson;
import io.swagger.client.model.mfn.PackageDimensions;
import io.swagger.client.model.mfn.UnitOfLength;
import io.swagger.client.model.mfn.UnitOfWeight;
import io.swagger.client.model.mfn.Weight;
import lambda.utils.MfnOrder;
import lambda.utils.MfnOrderItem;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.GetItemRequest;
import software.amazon.awssdk.services.dynamodb.model.GetItemResponse;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import static java.lang.Long.parseLong;
import static lambda.utils.Constants.INVENTORY_TABLE_HASH_KEY_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_HEIGTH_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_LENGTH_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_NAME_ENV_VARIABLE;
import static lambda.utils.Constants.INVENTORY_TABLE_SIZE_UNIT_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_STOCK_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_WEIGHT_UNIT_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_WEIGHT_VALUE_ATTRIBUTE_NAME;
import static lambda.utils.Constants.INVENTORY_TABLE_WIDTH_ATTRIBUTE_NAME;

public class InventoryCheckHandler implements RequestHandler<MfnOrder, MfnOrder> {

    public MfnOrder handleRequest(MfnOrder input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("InventoryCheck Lambda input: " + new Gson().toJson(input));

        long packageWeightValue = 0;
        String packageWeightUnit = "";

        long packageLength = 0;
        long packageWidth = 0;
        long packageHeight = 0;
        String packageSizeUnit = "";

        //Iterate over all order items and retrieve stock, size and weight from the database
        for (MfnOrderItem orderItem : input.getOrderItems()) {
            //Retrieve the item from DynamoDB by SKU
            //Update this section to match your product's logic
            Map<String, AttributeValue> key = new HashMap<>();
            key.put(INVENTORY_TABLE_HASH_KEY_NAME, AttributeValue.fromS(orderItem.getSku()));

            GetItemRequest getItemRequest = GetItemRequest.builder()
                    .tableName(System.getenv(INVENTORY_TABLE_NAME_ENV_VARIABLE))
                    .key(key)
                    .build();

            DynamoDbClient dynamoDB = DynamoDbClient.builder().build();
            GetItemResponse getItemResult = dynamoDB.getItem(getItemRequest);
            Map<String, AttributeValue> item = getItemResult.item();

            String stock = item.get(INVENTORY_TABLE_STOCK_ATTRIBUTE_NAME).n();
            if (parseLong(stock) < orderItem.getQuantity()) {
                throw new InternalError(
                        String.format("Stock level for SKU {%s} is not enough to fulfill the requested quantity",
                        orderItem.getSku()));
            }

            long itemWeightValue = parseLong(item.get(INVENTORY_TABLE_WEIGHT_VALUE_ATTRIBUTE_NAME).n());
            //Valid values for the database records are uppercase: [OZ, G]
            String itemWeightUnit = item.get(INVENTORY_TABLE_WEIGHT_UNIT_ATTRIBUTE_NAME).s();

            long itemLength = parseLong(item.get(INVENTORY_TABLE_LENGTH_ATTRIBUTE_NAME).n());
            long itemWidth = parseLong(item.get(INVENTORY_TABLE_WIDTH_ATTRIBUTE_NAME).n());
            long itemHeight = parseLong(item.get(INVENTORY_TABLE_HEIGTH_ATTRIBUTE_NAME).n());
            //Valid values for the database records are uppercase: [INCHES, CENTIMETERS]
            String itemSizeUnit = item.get(INVENTORY_TABLE_SIZE_UNIT_ATTRIBUTE_NAME).s();

            Weight itemWeight = new Weight();
            itemWeight.setValue(BigDecimal.valueOf(itemWeightValue));
            itemWeight.setUnit(UnitOfWeight.valueOf(itemWeightUnit));
            orderItem.setItemWeight(itemWeight);

            //Package weight is calculated by adding the individual weights
            //Update this section to match your selling partners' logic
            packageWeightValue += itemWeightValue;
            packageWeightUnit = itemWeightUnit;

            //Package size is calculated by adding the individual sizes
            //Update this section to match your selling partners' logic
            packageLength += itemLength;
            packageWidth += itemWidth;
            packageHeight += itemHeight;
            packageSizeUnit = itemSizeUnit;
        }

        input.setWeight(new Weight()
                .value(BigDecimal.valueOf(packageWeightValue))
                .unit(UnitOfWeight.valueOf(packageWeightUnit)));

        input.setPackageDimensions(new PackageDimensions()
                .length(BigDecimal.valueOf(packageLength))
                .width(BigDecimal.valueOf(packageWidth))
                .height(BigDecimal.valueOf(packageHeight))
                .unit(UnitOfLength.valueOf(packageSizeUnit)));

        return input;
    }
}

File: use-cases/shipping-v2/code/python/src/inventory_check_handler.py
import os
import boto3
import logging
from dataclasses import asdict
from src.utils import constants
from src.utils.shipping_utils import ShippingLambdaInput, Weight, Dimensions

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, context):
    # Log the input event for the Lambda
    logger.info(f"InventoryCheck Lambda input: {event}")

    # Initialize package weight and dimensions
    package_weight_value = 0
    package_weight_unit = "LB"

    package_length = 0
    package_width = 0
    package_height = 0
    package_size_unit = "GM"

    get_order_lambda_input = ShippingLambdaInput(**event)

    # Check if mfnOrder is not None
    if get_order_lambda_input.mfnOrder:
        # Check if orderItems is not None
        if get_order_lambda_input.mfnOrder.orderItems:
            # Iterate over all order items and retrieve stock, size, and weight from the database
            for order_item in get_order_lambda_input.mfnOrder.orderItems:
                # Retrieve the item from DynamoDB by SKU
                # Update this section to match your product's logic
                key = {constants.INVENTORY_TABLE_HASH_KEY_NAME: {"S": order_item.sku}}

                dynamodb = boto3.client(constants.AWS_DYNAMO_DB_CLIENT_KEY_NAME)
                get_item_result = dynamodb.get_item(TableName=os.environ.get(constants.INVENTORY_TABLE_NAME_ENV_VARIABLE),
                                                    Key=key)
                item = get_item_result.get("Item", {})

                stock = int(item.get(constants.INVENTORY_TABLE_STOCK_ATTRIBUTE_NAME, {"N": "0"})["N"])
                if stock < order_item.quantity:
                    raise Exception(f"Stock level for SKU {order_item.sku} "
                                    f"is not enough to fulfill the requested quantity")

                item_weight_value = int(item.get(constants.INVENTORY_TABLE_WEIGHT_VALUE_ATTRIBUTE_NAME, {"N": "0"})["N"])

                # Valid values for the database records are uppercase: [OZ, G]
                item_weight_unit = item.get(constants.INVENTORY_TABLE_WEIGHT_UNIT_ATTRIBUTE_NAME, {"S": ""})["S"]

                item_length = int(item.get(constants.INVENTORY_TABLE_LENGTH_ATTRIBUTE_NAME, {"N": "0"})["N"])
                item_width = int(item.get(constants.INVENTORY_TABLE_WIDTH_ATTRIBUTE_NAME, {"N": "0"})["N"])
                item_height = int(item.get(constants.INVENTORY_TABLE_HEIGHT_ATTRIBUTE_NAME, {"N": "0"})["N"])

                # Valid values for the database records are uppercase: [INCHES, CENTIMETERS]
                item_size_unit = item.get(constants.INVENTORY_TABLE_SIZE_UNIT_ATTRIBUTE_NAME, {"S": ""})["S"]

                # Create a Dimensions object for the item weight
                item_dimensions = Dimensions(unit=item_size_unit, length=item_length, width=item_width,
                                             height=item_height)

                # Create a Weight object for the item weight
                item_weight = Weight(unit=item_weight_unit, value=float(str(item_weight_value)))

                # Update the order item with the retrieved weight
                order_item.itemWeight = item_weight
                order_item.dimensions = item_dimensions

                # Package weight is calculated by adding the individual weights
                # Update this section to match your selling partners' logic
                package_weight_value += item_weight_value
                package_weight_unit = item_weight_unit

                # Package size is calculated by adding the individual sizes
                # Update this section to match your selling partners' logic
                package_length += item_length
                package_width += item_width
                package_height += item_height
                package_size_unit = item_size_unit

            get_order_lambda_input.mfnOrder.weight = Weight(unit=package_weight_unit,
                                                            value=float(str(package_weight_value)))

            get_order_lambda_input.mfnOrder.dimensions = Dimensions(length=package_length, width=package_width,
                                                                    height=package_height, unit=package_size_unit)
            
    return asdict(get_order_lambda_input)
